source/material_model/averaging.cc

```diff
@@ -0,0 +1,412 @@
+    }
+
+    template <int dim>
+    bool
+    Averaging<dim>::
+    compressibility_depends_on (const NonlinearDependence::Dependence dependence) const
+    {
+      return base_model->compressibility_depends_on(dependence);
+    }
+
+    template <int dim>
+    bool
+    Averaging<dim>::
+    specific_heat_depends_on (const NonlinearDependence::Dependence dependence) const
+    {
+      return base_model->specific_heat_depends_on(dependence);
+    }
+
+    template <int dim>
+    bool
+    Averaging<dim>::
+    thermal_conductivity_depends_on (const NonlinearDependence::Dependence dependence) const
+    {
+      return base_model->thermal_conductivity_depends_on(dependence);
+    }
+
+    template <int dim>
+    bool
+    Averaging<dim>::
+    is_compressible () const
+    {
+      return base_model->is_compressible();
+    }
+
+    template <int dim>
+    double
+    Averaging<dim>::
+    reference_viscosity() const
+    {
+      return base_model->reference_viscosity();
+    }
+
+    template <int dim>
+    double
+    Averaging<dim>::
+    reference_density() const
+    {
+      return base_model->reference_density();
+    }
+  }
+}
+
+// explicit instantiations
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(Averaging,
+                                   "averaging",
+                                   "The ``averaging'' Material model applies an averaging of the quadrature points"
+                                   "to any of the other available material models. In other words, it "
+                                   "is a ``compositing material model''."
+                                   "Parameters related to the average model are read from a subsection "
+                                   "``Material model/Averaging''. "
+                                   "The user must specify a ``Base model'' from which material properties are "
+                                   "derived. Furthermore an averaging operation must be selected, where the "
+                                   "Choice should be from the list none|arithmetic average|harmonic average|"
+                                   "geometric average|pick largest|NWD arithmetic average|NWD harmonic average"
+                                   "|NWD geometric average. "
+                                   "NWD stands for Normalized Weighed Distance. The models with this in front "
+                                   "of their name work by spanning a bell shaped curve around each quadrature "
+                                   "point which determines the weight of each quadrature point for the quadrature "
+                                   "point in question. Therefore the weight is determined by the distance between the "
+                                   "quadrature points. This distance is normalized to one, so the largest distance "
+                                   "becomes one."
```

Comment:
I think a formula that explains the averaging operation would be very helpful here. I don't think I can follow the text description.